File: services/core/src/simorgh_core/devices/observation_refresh_protocol.py
from __future__ import annotations

import time
from typing import Literal
from uuid import UUID, uuid4

from pydantic import BaseModel, ConfigDict, Field, model_validator

from simorgh_core.devices.protocol import PROTOCOL_VERSION, ProtocolEnvelope

OBSERVATION_REFRESH_CAPABILITY = "android.observation.refresh.v1"
OBSERVATION_REFRESH_REQUEST_TYPE = "device.observation_refresh"
OBSERVATION_REFRESH_ACK_TYPE = "device.observation_refresh_ack"

ObservationRefreshAckStatus = Literal[
    "accepted",
    "duplicate",
    "busy",
    "expired",
    "observer_unavailable",
    "rejected",
]


class DeviceObservationRefreshPayload(BaseModel):
    model_config = ConfigDict(extra="forbid")

    request_id: UUID
    timeout_ms: int = Field(ge=250, le=10_000)
    expected_state_fingerprint: str | None = Field(
        default=None,
        min_length=64,
        max_length=64,
        pattern=r"^[0-9a-f]{64}$",
    )
    expected_active_package: str | None = Field(
        default=None,
        min_length=1,
        max_length=512,
    )
    reason: str = Field(default="operator requested fresh observation", max_length=1_000)


class DeviceObservationRefreshAckPayload(BaseModel):
    model_config = ConfigDict(extra="forbid")

    request_id: UUID
    status: ObservationRefreshAckStatus
    received_at_ms: int = Field(ge=0)
    detail: str = Field(default="", max_length=1_000)


class DeviceObservationRefreshEnvelope(ProtocolEnvelope):
    type: Literal["device.observation_refresh"] = OBSERVATION_REFRESH_REQUEST_TYPE

    @classmethod
    def create(
        cls,
        *,
        device_id: UUID,
        payload: DeviceObservationRefreshPayload,
        message_id: UUID | None = None,
    ) -> DeviceObservationRefreshEnvelope:
        request_id = message_id or payload.request_id
        if payload.request_id != request_id:
            raise ValueError("refresh payload request_id must equal envelope message_id")
        return cls(
            protocol_version=PROTOCOL_VERSION,
            message_id=request_id,
            sent_at_ms=int(time.time() * 1000),
            device_id=device_id,
            payload=payload.model_dump(mode="json"),
        )

    @model_validator(mode="after")
    def validate_request_identity(self) -> DeviceObservationRefreshEnvelope:
        payload = DeviceObservationRefreshPayload.model_validate(self.payload)
        if payload.request_id != self.message_id:
            raise ValueError("refresh request_id must equal envelope message_id")
        if self.correlation_id is not None:
            raise ValueError("refresh request envelope cannot declare correlation_id")
        return self


class DeviceObservationRefreshAckEnvelope(ProtocolEnvelope):
    type: Literal["device.observation_refresh_ack"] = OBSERVATION_REFRESH_ACK_TYPE

    @classmethod
    def create(
        cls,
        *,
        device_id: UUID,
        request_envelope_id: UUID,
        payload: DeviceObservationRefreshAckPayload,
    ) -> DeviceObservationRefreshAckEnvelope:
        if payload.request_id != request_envelope_id:
            raise ValueError("refresh ACK request_id must equal correlation_id")
        return cls(
            protocol_version=PROTOCOL_VERSION,
            message_id=uuid4(),
            sent_at_ms=int(time.time() * 1000),
            device_id=device_id,
            correlation_id=request_envelope_id,
            payload=payload.model_dump(mode="json"),
        )

    @model_validator(mode="after")
    def validate_ack_identity(self) -> DeviceObservationRefreshAckEnvelope:
        payload = DeviceObservationRefreshAckPayload.model_validate(self.payload)
        if self.correlation_id is None:
            raise ValueError("refresh ACK requires correlation_id")
        if payload.request_id != self.correlation_id:
            raise ValueError("refresh ACK request_id must equal correlation_id")
        return self

File: apps/android/src/main/java/ai/simorgh/android/protocol/ObservationRefreshProtocol.kt
package ai.simorgh.android.protocol

import ai.simorgh.android.accessibility.AccessibilitySnapshot
import kotlinx.serialization.SerialName
import kotlinx.serialization.Serializable
import kotlinx.serialization.json.encodeToJsonElement
import kotlinx.serialization.json.jsonObject
import java.util.UUID

@Serializable
data class DeviceObservationRefreshPayload(
    @SerialName("request_id")
    val requestId: String,
    @SerialName("timeout_ms")
    val timeoutMs: Long,
    @SerialName("expected_state_fingerprint")
    val expectedStateFingerprint: String? = null,
    @SerialName("expected_active_package")
    val expectedActivePackage: String? = null,
    val reason: String = "operator requested fresh observation",
)

@Serializable
enum class ObservationRefreshAckStatus {
    @SerialName("accepted")
    ACCEPTED,

    @SerialName("duplicate")
    DUPLICATE,

    @SerialName("busy")
    BUSY,

    @SerialName("expired")
    EXPIRED,

    @SerialName("observer_unavailable")
    OBSERVER_UNAVAILABLE,

    @SerialName("rejected")
    REJECTED,
}

@Serializable
data class DeviceObservationRefreshAckPayload(
    @SerialName("request_id")
    val requestId: String,
    val status: ObservationRefreshAckStatus,
    @SerialName("received_at_ms")
    val receivedAtMs: Long,
    val detail: String = "",
)

object ObservationRefreshProtocol {
    const val TYPE_REFRESH: String = "device.observation_refresh"
    const val TYPE_REFRESH_ACK: String = "device.observation_refresh_ack"
    const val CAPABILITY: String = "android.observation.refresh.v1"

    fun decodeRequest(envelope: ProtocolEnvelope): DeviceObservationRefreshPayload {
        require(envelope.type == TYPE_REFRESH) { "unexpected refresh message type" }
        require(envelope.correlationId == null) {
            "refresh request cannot declare correlation_id"
        }
        val payload = DeviceProtocol.json.decodeFromJsonElement(
            DeviceObservationRefreshPayload.serializer(),
            envelope.payload,
        )
        return validateRequest(
            requestEnvelopeId = envelope.messageId,
            payload = payload,
        )
    }

    fun acknowledgement(
        deviceId: String,
        requestEnvelopeId: String,
        requestId: String,
        status: ObservationRefreshAckStatus,
        detail: String = "",
        nowMs: Long = System.currentTimeMillis(),
    ): ProtocolEnvelope {
        validateRequestIdentity(requestEnvelopeId, requestId)
        return ProtocolEnvelope(
            messageId = UUID.randomUUID().toString(),
            type = TYPE_REFRESH_ACK,
            sentAtMs = nowMs,
            deviceId = deviceId,
            correlationId = requestEnvelopeId,
            payload = DeviceProtocol.json.encodeToJsonElement(
                DeviceObservationRefreshAckPayload(
                    requestId = requestId,
                    status = status,
                    receivedAtMs = nowMs,
                    detail = detail.take(1_000),
                ),
            ).jsonObject,
        )
    }

    fun correlatedObservation(
        deviceId: String,
        refreshEnvelopeId: String,
        streamId: String,
        sequence: Long,
        stateFingerprint: String,
        snapshot: AccessibilitySnapshot,
        messageId: String = UUID.randomUUID().toString(),
        nowMs: Long = System.currentTimeMillis(),
    ): ProtocolEnvelope {
        requireUuid(refreshEnvelopeId, "refresh envelope id")
        return ProtocolEnvelope(
            messageId = messageId,
            type = DeviceProtocol.TYPE_OBSERVATION,
            sentAtMs = nowMs,
            deviceId = deviceId,
            correlationId = refreshEnvelopeId,
            payload = DeviceProtocol.json.encodeToJsonElement(
                DeviceObservationPayload(
                    streamId = streamId,
                    sequence = sequence,
                    stateFingerprint = stateFingerprint,
                    snapshot = snapshot,
                ),
            ).jsonObject,
        )
    }

    fun validateRequest(
        requestEnvelopeId: String,
        payload: DeviceObservationRefreshPayload,
    ): DeviceObservationRefreshPayload {
        validateRequestIdentity(requestEnvelopeId, payload.requestId)
        require(payload.timeoutMs in 250..10_000) {
            "refresh timeout_ms must be in 250..10000"
        }
        payload.expectedStateFingerprint?.let { fingerprint ->
            require(fingerprint.length == 64 && fingerprint.all(::isLowercaseHex)) {
                "expected_state_fingerprint must be 64 lowercase hex characters"
            }
        }
        payload.expectedActivePackage?.let { packageName ->
            require(packageName.isNotBlank() && packageName.length <= 512) {
                "expected_active_package is invalid"
            }
        }
        require(payload.reason.length <= 1_000) { "refresh reason exceeds 1000 characters" }
        return payload
    }

    private fun validateRequestIdentity(requestEnvelopeId: String, requestId: String) {
        requireUuid(requestEnvelopeId, "refresh envelope message_id")
        require(requestId == requestEnvelopeId) {
            "refresh request_id must equal envelope message_id"
        }
    }

    private fun requireUuid(value: String, field: String) {
        require(runCatching { UUID.fromString(value) }.isSuccess) { "$field must be a UUID" }
    }

    private fun isLowercaseHex(character: Char): Boolean =
        character in '0'..'9' || character in 'a'..'f'
}
